packages/selfcare-client-users-updater/src/services/selfcareClientUsersUpdaterProcessor.ts

```diff
@@ -0,0 +1,140 @@
+import {
+  getAllFromPaginated,
+  getInteropHeaders,
+  logger,
+  RefreshableInteropToken,
+  userRole,
+} from "pagopa-interop-commons";
+import { EachMessagePayload } from "kafkajs";
+import {
+  generateId,
+  CorrelationId,
+  genericInternalError,
+} from "pagopa-interop-models";
+import { bffApi } from "pagopa-interop-api-clients";
+import { match, P } from "ts-pattern";
+import { AuthorizationProcessClient } from "../clients/authorizationProcessClient.js";
+import {
+  selfcareUserEventType,
+  UsersEventPayload,
+  relationshipStatus,
+} from "../model/UsersEventPayload.js";
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function selfcareClientUsersUpdaterProcessorBuilder(
+  refreshableToken: RefreshableInteropToken,
+  authorizationProcessClient: AuthorizationProcessClient,
+  productId: string
+) {
+  return {
+    async processMessage({
+      message,
+      partition,
+    }: EachMessagePayload): Promise<void> {
+      const correlationId: CorrelationId = generateId();
+
+      const loggerInstance = logger({
+        serviceName: "selfcare-client-users-updater",
+        correlationId,
+      });
+
+      try {
+        loggerInstance.info(
+          `Consuming message for partition ${partition} with offset ${message.offset}`
+        );
+
+        if (!message.value) {
+          loggerInstance.warn(
+            `Empty message for partition ${partition} with offset ${message.offset}`
+          );
+          return;
+        }
+
+        const stringPayload = message.value.toString();
+        const jsonPayload = JSON.parse(stringPayload);
+
+        if (jsonPayload.productId !== productId) {
```

Comment:
Here, you could directly parse using the zod model, instead of doing JSON.parse and then parse using the model

```suggestion
        const userEventPayload = UsersEventPayload.parse(stringPayload);

        if (userEventPayload.productId !== productId) {
```

Reply:
In this case this check could be included in the pattern match @Alepazz

Reply:
@ecamellini I don't think Zod's parser can convert a string directly — it expects a TypeScript object. That's why we need to parse the JSON first. Does that make sense to you?

Reply:
True that, ignore my previous comment :))